### ISSUE TITLE
Integrating LLMs into search (link prediction, top-site summarization, stable diffusion images, academic articles)

### ISSUE THREAD

**Issue by alexlyee** [searxng/searxng#2163]
There are plenty of amazing solutions for using large language models (LLMs) to help with searching. For sake of compressing this request, I'll point out four kinds of them that I want in a modern search engine:
- [Metaphor systems](https://metaphor.systems) has a working beta which is trained to _predict links based on prompts._ This is much more like a classic search engine, except, my god, it already has been so much better at getting useful articles than any single search engine has been for my searches.
- [Perplexity](https://www.perplexity.ai/), [KOMO](https://komo.ai/), and [Uminal](https://www.uminal.com/) have created chat bots that take any text sent to them, forward it to a search engine, pick the top x resulting sites, and feed them through an LLM that summarizes them. They all have been fantastic for cutting through the clutter that I'm so used to having to sift through on the internet and getting exactly the answer I want. The first two include citations, links to the sites where they got the facts in their answers from, which is amazing. Uminal may be the most useful to us though, as I'm aware the owner is considering open-sourcing it (!)
- Though I know there are many, [Lexica](https://lexica.art) is the longest-standing AI image search engine I know of.
- [Elicit](https://elicit.org/) is the best solution I know of today for searching scholarly articles using AI, and it's fantastic.

It would be great if we could implement a menu on the side for link-prediction and top-site summarizing LLMs like those outlined. Perhaps stable-diffusion image and AI-based academic article searchers could be implemented as engines, or perhaps they ought to have their own tabs?

If there isn't an API available, is it a possibility to have a headless browser scrape a response page?

Maybe it would be a good idea to cache responses to eliminate unnecessary duplicate requests to these aforementioned free providers?

I'm aware of #2009, but I don't think with these offerings that already exist we should use ChatGPT to answer prompts directly, though that is certainly an option (!)

I'd be happy to work on this here if we think that this is the way forward. Please give input if you know of better ways of achieving this right now!

cheers,
alex

----

Related:

- https://github.com/searxng/searxng/issues/2008
- https://github.com/searxng/searxng/issues/2273

**Comment by blob42** [comment]
This might be slightly related to #2129 .

 I am currently adding a PR to [langchain](https://github.com/hwchase17/langchain/pull/854) that would allow the usage of SearxNG API to be used as a search engine for LLMs. 

If SearX is to be used with other LLM frameworks or libraries, for a quick fix  I needed to allow whitelising for the rate limiter for those who self host instances. A proper solution is to allow unlimited usage of the API using a key.

**Comment by blob42** [comment]
@alexlyee 

Follow up to my previous comment to add more context and reply to the original topic of this issue. 

First, I believe integrating smart answers/prompting would really make Searx shine as an alternative free(libre) meta search engine. As a searx user (self hosted), my main reason is privacy and control of my search results. 

Prompt/LLM based search assistants sounds amazing but right now the main problem is that for each tool/assistant I am required to login with Google/Discord ... and provide my private data to be harvested. Given the unlimited potential of LLMs for enhanced information retrieval, at this rhythm we will have a ridiculous number of tools, each one requiring signup or some sort of API key to use. 

Hence, my interest for using langchain with searx. Langchain is a toolkit to build LLM based smart agents, chat bots etc. It allows you to quickly hook LLM chain calls to the internet or other APIs. Almost all the tools that where mentioned as examples could be reproduced with minimal code using Langchain.

In order to use LLMs langchain made an abstract api to hook in any LLM, right now I believe it's mainly OpenAI, but they will be able to handle other LLMs including free and open source ones in the future. 

In summary, SearX could offer LLM assisted search by developing its own modules that would use Searx itself for searching. A typical use case would be to allow users to input their OpenAI API key and enable AI based search. A big advantage as well, langchain is a python project and so is searx, it would be easy to quickly come up with equivalents to many of the proprietary tools that are proliferating at an insane pace.

**Comment by alexlyee** [comment]
@blob42 you bring up an excellent point that currently openAI requires a phone number to sign up which can be used to identify you. It's not easy (and it's cost-prohibitive) to obtain an anonymized VoIP number for use with openAI. The other services I mentioned are usable with just an email, but I wonder if it will stay that way.

You make several other great points, like the fact that they are both Python based, however, I don't think it's necessary to interact with OpenAI at this point now that tools like metaphor.systems and perplexity exist.

**Comment by blob42** [comment]
@alexlyee using langchain does not require OpenAI, they already adapted the hugging face library as an LLM and many others as well. The LLM part is only a small piece that is required to build more advanced assistants. Say in a year or two  [OpenAssistant](https://github.com/LAION-AI/Open-Assistant) is ready it will be trivial to add to LangChain. 

My point being, instead of integrating specific tools with Searx, it would be great to integreate a framework that would allow anyone to build their own assistant experience the same way it is done with `engines`. On the searx side an API needs to be defined for developping assistants or agents, then they can be activated or not in the settings or UI. 

To use some of your examples, let's pick [Metaphor](https://metaphor.systems/). In langchain a similar can be made in many ways, for example using  [ReAct](https://langchain.readthedocs.io/en/latest/modules/agents/implementations/react.html). 

I will find some time soon to make a quick notebook demonstrating the feature using langchain and searx results.

**Comment by alexlyee** [comment]
@blob42 I see, and yeah I was thinking the same when I found out about OpenAssistant only a few days ago. That sounds excellent! I agree that it would be better to go about it that way. I suppose I figured that it would be quickest to go about it by using the aforementioned already built services, and that would come later. I've got to learn more about langchain, it seems to be something really good to know about. Don't know where to start.
Also I noticed that someone took the time to 👎 this topic-- is there something we've got really wrong here?

**Comment by blob42** [comment]
We could create an MVP with some simple LLM assistant. My proposal would be to split the work in these steps:

1. Prototype some assistant scenarios in notebooks, it will be easy when my Searx wrapper for langchain will be merged. 
2. Draft some ideas for an API / conventions for adding `assistants` the same way as `engines`.
3. What would the UI side look like ? I would consider adding them in a different view than the default one so the Searx experience stays exactly the same for those who are not interested in this. 

In the meantime let's wait for some feedback from other users and core maintainers.

**Comment by alexlyee** [comment]
Agreed. I think the link-predicting models like metaphor should be made as an engine, while all the others that return anything more than links should be classified as assistants. Like I mentioned before, the assistants could be placed on the right hand side of the search results in their own boxes. Do you think it would be best to have a chat box in there or a link to a page where one can continue the conversation in a place not unlike ChatGPT's UI?

**Comment by bgvaughan** [comment]
99% of the reason I set up a SearXNG instance for my personal use is that I was hoping to filter out LLM-generated nonsense. I would consider deliberately including it grossly unethical.

**Comment by alexlyee** [comment]
@bgvaughan thank you for your opinion, I was wondering what the stance was of the people downvoting! I guess I can understand that, although I don't see the harm in making it an option. I wouldn't have opened this suggestion if I didn't find tools like metaphor.systems extremely useful myself. Since I made that initial post, probably a quarter of my searches online have been made using that tool, and it's been far more effective than a regular search engine for certain things. Massive productivity booster for me, and so I'd love to have it somehow integrated into my every search, and I'm sure I'm not alone in that opinion

**Comment by dinvlad** [comment]
Tbh while I remain skeptical about LLMs too, I agree with @alexlyee this should at least be an opt-in. As these technologies mature, while they will likely never be perfect, they can at least be useful to some. Like @blob42, I also think that LangChain goes a long way towards preserving privacy, for those like myself who might prefer to have a locally-hosted option.

**Comment by rasos** [comment]
With Mistral's open sourced 8x7B model there is a new powerful LLM which one could even host locally https://mistral.ai/news/mixtral-of-experts/ 

Agree that results from any LLM should be opt-in and always clearly indicated as AI generated.

**Comment by dalf** [comment]
FYI [Perplexica](https://github.com/ItzCrazyKns/Perplexica) is the other way around: it relies on SearXNG among other projects to provide an AI assistant.

**Comment by nemobis** [comment]
LLMs are being [forced down the throat](https://pivot-to-ai.com/2025/02/06/how-to-remove-copilot-ai-from-your-office-365-subscription-hit-cancel/) of billions already, coming from all directions. There's asymptotically zero need for yet another provider.

**Comment by puresick** [comment]
https://github.com/open-webui/open-webui also allows using SearXNG as search provider for LLMs.

**Comment by rubin110** [comment]
A lot has changed in the world since the creation of this feature request. I'm providing my feedback here, from someone whose professional expertises covers understanding the user's approach to technology (or the lack there of). Disclaimer, I'm not an expert on LLMs and am still running into new projects that offer different local capabilities of interacting with them...

With the growing intensity of bad state actors and their gross government overreach (regardless of where you live in the world at this point), there's been a resurgence of internet citizens (of all skill levels) who want to go the route of a more privacy focused life when handling their data and habits online. I've started a _self-hosting at home for newbies_ meetup at [my local hacker space](https://pumpingstationone.org/) to help others who are just starting to pick up the skills needed to set up tools, like SearXNG, and VPNs, and Home Assistant, and Jellyfin or Plex, or even just Linux on a Raspberry Pi. For most of these people their decision to switch from a (free) commercial product to an open source project pivots on how much functionality can be replicated by the replacement tool.

There's also the adoption of the tools by people who don't actually end up setting it up, such as housemates, friends, significant others, or members of our local non-profit/communities/hacker space. Even though there's a desire to use tools to mitigate their online footprint, it's a coin toss if it'll stick when the tool doesn't work and function just like its commercial counterparts.

Will the creation of an option for an admin to link to a local (or publically shared) LLM in order to populate the top of the search results page with some human-readable summary of the answer to whatever the user is searching for...
- Is needed to make SearXNG useful? No.
- Helps SearXNG be more familiar to non-techies users, promoting more adoption? Yes.

I think that any addition to SearXNG that gets more people onboard is worth investigating time into (and yes that's being shared from someone who doesn't contribute code to this project).

That being said, in my mind implementing something that mimics the current behaviors of AI summaries provided at the top of result pages, and tying it into auto-completion (ignoring images and academic articles for now) would suck. When reviewing some use cases in my head, here are some sucky thoughts...
- The debate of utilizing a public LLM API and how un-private that is can go on forever.
- Needing (in most cases) to pay for API access to said public LLMs also puts a damper on setting up this feature.
- If going the self-hosted route, I just spent a month setting up Ollama locally and it was painful. Lots of learning to do things along the way that isn't described clearly, strewn about dozens of terribly AI written Medium articles and YouTube videos (we have truly lost the internet). Setting this stuff up isn't trivial.
- Finding a model that works well for your needs is like trying out all the pizza restaurants in your city and then deciding which one you like the best. It's a bit of work.
- The large more useful models are expensive to run at home as you basically need a GPU with a thick stack of VRAM.
- Smaller edge LLMs that can run on a bare CPU and system RAM are less useful (in my humble opinion).
- The real world data that downloadable models contain are dated, many back to 2022-2023. So some back and forth interaction must be completed between the search engine and the LLM to decide if actual links to the search results need to be open in order for the LLM to gain more recent knoweldge to provide an appropriate answer, which is slow. "What's been the impact of climate change on the trade routes between SE Asia and Scandinavia" or "Why did the US Democratic Party lose the 2025 presidential elections?"
- LLMs could make drastically different assumption about the users search query intention than the aggregated engines. "Why did the conservatives win the last election?" vs "Why did the conservatives win the last **US** election?" So answers might not line up with the actual results in the page.
- [All LLMs have biases](https://www.theguardian.com/technology/2025/jan/28/we-tried-out-deepseek-it-works-well-until-we-asked-it-about-tiananmen-square-and-taiwan).

I could spend a day writing up some other terrible use cases here. It's a lot. I think many of these have gotten speced and solved by other available tools (open source community editions of paid for SaaS solutions).

I do have Open WebUI setup to pull search results from my instance of SearXNG. It's good if you've got a decent model setup, but again slow...

https://github.com/user-attachments/assets/c0c784b2-68bf-44db-9569-b19c79fc1579

Open WebUI does have an API available to make calls to it for such queries

I will say that while writing this I stumbled onto [Duck.ai](https://duckduckgo.com/duckduckgo-help-pages/duckai/approach-to-ai/) and DuckDuckGo's implementation of the AI response on their results page, which I think is fantastic from a privacy perspective. The user has the option **after the page has fully loaded**, to opt-in to sending a search query prompt to their LLM and receiving a summary.

https://github.com/user-attachments/assets/101b2bc9-c2be-49fd-b480-ae35d336368a

Duck.ai runs its own LLM service, which is also nice. If I've already got DuckDuckGo set as my auto completer in SearXNG, it would be neat to have a similar option to enable this style of opt-in AI summary. They do have a daily limit to the number of responses, and sadly they don't have an API available yet.

Anyhow, at the end of writing this I still think that eventually (soon) SearXNG will need to offer some means for LLMs to provide summaries and predictions in order to continue its appeal to new users. But I honestly I don't know if your project has a goal or mission it's trying to fulfill, and if drawing in users from all skill levels and sides of the internet is part of it.

If you've gotten this far, thanks for reading all of this! And appreciation to all the contributors, I'm really happy I've gotten this setup recently, as is my partner.

**Comment by bgvaughan** [comment]
I appreciate the thoroughness of @rubin110 's comment. My principle objection to LLM-generated summaries in search tools is ethical: LLMs generate texts that appear to be analyses by a reasonable and responsible person, but they are not. This is misleading. By presenting LLM-generated text as a valid search result, that risks re-enforcing that misleading presentation.

**Comment by rubin110** [comment]
Trying my best to not make this sound like a debate, just providing a different perspective.

@bgvaughan I agree with you on that the factual accuracy of what's returned by an LLM _sometimes_ isn't perfect, and how it is written does make it seem like a human wrote it.

However, I don't believe that how well human-like something is written can be used as a marker for its factual accuracy.

Wikipedia is plugged with inaccuracy that's reinforced by the [oligarchy](https://web.archive.org/web/20250115172424/https://www.telegraph.co.uk/technology/2023/04/28/wikipedia-elite-control-the-worlds-knowledge/) of [high level editors](https://www.theguardian.com/technology/2009/aug/12/wikipedia-deletionist-inclusionist), written by genuine human beings.

On the daily we see articles/blog posts/how to guides that were written by AI prompts and then have the "author" slap a name and random profile photo on it and SEO the hell out of it, which no real checks on if the content is legit. All of these being the first couple of items from our search results. And then there's AI on top of all that by the major search engines to try to figure out what search results are actually valid and which ones are sort of kind of maybe incorrect.

There's also cultural, religious, geopolitical, and scientific bias when you start to dive into fact checking. 

Is AI contributing to littering the internet of misinformation? 100% totally true. Would not providing an AI summarized result on the right column prevent the user from gaining inaccurate information? No, because misinformation is already being disseminated in the search results, and it all looks legitimate. In fact (in my opinion) more so misleading than an AI response with a disclaimer attached to the bottom of it.

If SearXNG's goal is to prevent misleading information from being delivered to the user, then the project needs its own web crawler and fact checking system, and even then how do you make that unbiased?

Edit: **What is SearXNG's goal?**

**Comment by return42** [comment]
Thanks for your https://github.com/searxng/searxng/issues/2163#issuecomment-2675267531, it lists a lot of things that have been on my mind ... and questions that I don't have answers to either, which makes me hesitate to deal with LLM in SearXNG (and because there is so much other work to do ;-)

... I am therefore only taking some very general stances ...

> Even though there's a desire to use tools to mitigate their online footprint, it's a coin toss if it'll stick when the tool doesn't work and function just like its commercial counterparts.

:thinking: .. convenience has never been a guarantee of independence. *Those who have their food served to them will never know what they really like .. and they certainly won't determine the list of ingredients - if you want that, you have to reach for the wooden spoon yourself.*

> What is SearXNG's goal?

1. privacy
2. privacy
3. unbiased search results
4. comfort

Do we need AI in SearXNG? ... or do we already have it?

----

*attention: you read a private biased opnion!*

For me, the cost-benefit ratio is not yet good enough ... you have already mentioned the reasons for this. ... many summaries I've seen contain massive misinformation that only experts can recognize, but they don't need such a summary.

And those who are not subject matter experts (the majority of society) then have a misconception ... which boosts their confidence and at the same time their cluelessness ... a change that we can observe in the majority of society.

In some cases, the results are contradictory in the different national languages/regions ... which then leads to the division of the clueless, which we can also observe 

----

We have the "Answers" in SearXNG: these are usually the "Answers" that we take from search engines like DDG and also display at the top. Most of the time, summaries from Wikipedia articles are displayed there. But sometimes AI-generated content is also shown.

Some of these (AI generated) "Answers" that we take over are so bad that I have already thought about how to separate the wheat from the chaff. But that is not possible. Not without enormous technical effort (AI to eliminate AI from our live 🤦) and there would always be a bias involved.

-----

As a maintainer of SearXNG, I would also like to make it clear:

1. we pass on research results 1:1, this should also apply to AI-generated content.

2. If we receive serious PRs on the topic, we will seriously address it in a review 👍 

.. if it is technically feasible and the goals of SearXNG are not violated, we will also integrate AI... just as we are prepared to integrate any calculator whose results we do not check individually ;-)

**Comment by rubin110** [comment]
@return42 Thanks for the awesome response!

**Comment by cra88y** [comment]
I am working on a plugin to add an async ai response using the answer box. It uses some of the results for grounding and it doesn’t block the normal results load time. Still some bugs to iron out but it works.

Supports OpenRouter, Gemini, Ollama, OpenAI APIs.

The default model set is “:free” on OpenRouter (1k req/day) and on free Gemini API for ~15k free requests per day. Gemma 27B has been fast enough for me (1-5s) and not as stupid as the even lighter models.

I haven’t seen any INSIDE searxng ai answer stuff yet so maybe someone would use it or help with it :)

https://github.com/cra88y/ai-answers-searxng

**Comment by Mavyre** [comment]
@cra88y Your plugin is amazing. I had to adapt it a bit to make it work with OpenWebui's Ollama proxy, but it is working like a charm. I guess it could be enhanced a bit by really scraping the top results content and adding a "following up" text box.

But it really seems to be an answer to this feature request!

**Comment by alexlyee** [comment]
@rubin110 100%. This is great because it shows that AI-generated text is often a continuation of already misleading information online, rather than a new kind of it. People already get most of their information many times removed from original sources—not sure LLMs are categorically different.

Will check out the plugin soon!

**Comment by rubin110** [comment]
Wow, a year later, I'm surprised/not surprised this has picked up again! I'm in the middle of a move but would be happy to do some QA after the dust settles in my life.

I will say that since I wrote everything above, I've sort of given up on running a local LLM. It's expensive, GPUs with massive VRAM don't grow on trees and the cost of power usage isn't something that's easy to ignore. I have dived head first into using Claude for the majority of my AI tasks. I understand that Anthropic is a for-profit company, and their privacy agreement is only as good as the paper it's written on. But they are the closest I can get with my desires of functionality and needs of privacy. And yes I know they've totally done some unethical things with data digestion.

This all feels the same as it did ~10 years ago where I initially began running a very broken instance of `exim4` on a raggedy-ass home server and Thunderbird on my hacker standard ThinkPad X220, periodically not noticing that I wasn't receiving email randomly because I messed up a config. To giving up and switching to Gmail because email was one thing that was hard to swim against the global current. I could still PGP encrypt messages as needed, and for my threat vectors at the time having email hosted by the company with the largest security team in the world covered all of my bases.

Professionally I operate QA teams for software/hardware companies. There's been a very observable push by people who don't know anything about engineering towards using AI for everything, especially testing automation, because a CEO's top goal is to make money for investors and cutting costs is what they do. "AI" and "LLM" are market buzz words now, of which the majority of people don't understand what they mean. I want this to be different, I fully believe manual testing by a human, before automation, and well before having an LLM come in a blindly do shitty code validation. However, I'm telling all of my other QA friends (and general software devs) that if they want employment they must pick up AI now as a **new tool to have under their belt**, or start looking for other different careers to switch over to. Sink or swim.

@bgvaughan I get you're against LLMs in a privacy focused tool. But LLMs are longer a simple polarized line to take sides over. We have to approach technologies like this with an eye for the nuanced, understand how others use it and their requirements. And then we must do our best to contribute back into the ongoing conversations with feedback that contain the essence of our concerns, but still compliments what's going on around us.

Sorry for getting philological there. The coffee just kicked in.

Edit: @bgvaughan I won't hold it against you if you decide to emoji thumbs down me. Everyone must have a voice.

**Comment by bgvaughan** [comment]
My thumbs down was in reference to an expression of extreme cynicism regarding human knowledge and culture. LLMs scramble text, and only accidentally produce text that resembles truth. To value the output of LLMs over text written by human beings is to deny the value of human subjectivity. It is a fascist world view, and I must oppose it.

**Comment by alexlyee** [comment]
@bgvaughan I think we're closer than you realize. My point was that corporatism and bad actors have already flooded the internet with misleading human-written content—that's the problem I was describing, not a dismissal of human knowledge. LLMs inherit that mess; they didn't create it.

**Comment by xanathon** [comment]
Since I use SearxNG explicitly to get away from hallucinating LLMs, this would be a deal breaker for me. I really hope, if something like this actually is implemented, it's all completely optional and has to be activated in settings instead of forced onto me like Big Tech does.

**Comment by Mavyre** [comment]
> Since I use SearxNG explicitly to get away from hallucinating LLMs, this would be a deal breaker for me. I really hope, if something like this actually is implemented, it's all completely optional and has to be activated in settings instead of forced onto me like Big Tech does.

Totally understandable. If it's included in Searxng, I think it would be as an optional (disabled by default plugin). In all the case, it requires either an external API key, or a server with a GPU to have such a feature, so it wouldn't be wise to be there by default.

**Comment by nutsnox** [comment]
Please implement some kind of AI search summary/chat feature in SearxNG that, minimally, can utilize Ollama/OpenAI compatible API.
